src/openloco/windows/stationwnd.cpp

```diff
@@ -62,12 +667,237 @@ namespace openloco::ui::windows::station
                     }
                 }
                 uint8_t amount = (rating * 327) / 256;
-                draw_rating_bar(w, dpi, 100, y, amount, colour);
+                draw_rating_bar(self, dpi, 100, y, amount, colour);
 
                 uint16_t percent = rating / 2;
-                gfx::draw_string_494B3F(dpi, 201, y, 0, string_ids::station_cargo_rating_percent, &percent);
+                gfx::draw_string_494B3F(*dpi, 201, y, 0, string_ids::station_cargo_rating_percent, &percent);
                 y += 10;
+                
+            }
+        }
+
+        static void initEvents()
+        {
+            events.draw = draw;
+            events.on_mouse_up = on_mouse_up;
+            events.on_resize = on_resize;
+            events.on_update = common::update;
+            events.prepare_draw = prepare_draw;
+            events.text_input = common::text_input;
+            events.get_scroll_size = get_scroll_size;
+            events.tooltip = tooltip;
+            events.draw_scroll = draw_scroll;
+        }
+    }
+
+    namespace common
+    {
+        struct TabInformation
+        {
+            widget_t* widgets;
+            const widx widgetIndex;
+            window_event_list* events;
+            const uint64_t* enabledWidgets;
+        };
+
+        static TabInformation tabInformationByTabOffset[] = {
+            { station::widgets, widx::tab_station, &station::events, &station::enabledWidgets },
+            { cargo::widgets, widx::tab_cargo, &cargo::events, &cargo::enabledWidgets },
+            { cargo_ratings::widgets, widx::tab_cargo_ratings, &cargo_ratings::events, &common::enabledWidgets }
+        };
+
+        static void prepare_draw(window* self)
```

Comment:
Please generally add a comment what function on what address this implementation is based. Here, I believe:
```suggestion
        // 0x0048EC3B
        static void prepare_draw(window* self)
```

Reply:
It's based off the identical part of three seperate subs 0x48E352, 0x48E7C0 and 0x48EC3B.
I'll list them all.

Reply:
Ah, sorry, I hadn't noticed it was a common sub. In that case, it's not necessary to list the addresses.